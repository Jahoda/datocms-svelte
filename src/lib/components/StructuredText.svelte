<script context="module" lang="ts">
	export type { StructuredText, Document } from 'datocms-structured-text-utils';
</script>

<script lang="ts">
	import type { SvelteComponentTyped } from 'svelte';
	import type { StructuredText, Node as N } from 'datocms-structured-text-utils';

	import Node from './_StructuredText/Node.svelte';

	/** The actual field value you get from DatoCMS **/
	// FIXME: add support for documents (from CMA) and node (for rendering branches)
	export let data: StructuredText | null = null;

	type PredicateComponentTuple = [(n: N) => boolean, new (...any: any[]) => SvelteComponentTyped];

	export let components: PredicateComponentTuple[] = [];

	$: node = data?.value.document;
	$: blocks = data?.blocks;
	$: links = data?.links;
</script>

{#if node}
	<Node {node} {blocks} {links} {components} />
{/if}
