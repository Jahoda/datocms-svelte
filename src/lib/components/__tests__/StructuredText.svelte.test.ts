import { describe, expect, it } from 'vitest';
import '@testing-library/jest-dom';
import { render } from '@testing-library/svelte';

import {
	RenderError,
	isBlock,
	isHeading,
	isInlineItem,
	isItemLink,
	isSpan
} from 'datocms-structured-text-utils';

import StructuredText, { type StructuredText as ST } from '../StructuredText.svelte';

import CustomSpan from './__fixtures__/CustomSpan.svelte';
import IncreasedLevelHeading from './__fixtures__/IncreasedLevelHeading.svelte';
import ItemLink from './__fixtures__/ItemLink.svelte';
import Block from './__fixtures__/Block.svelte';
import InlineItem from './__fixtures__/InlineItem.svelte';
import type { DocPageRecord, QuoteRecord } from './__fixtures__/types';

describe('StructuredText', () => {
	describe('with no value', () => {
		it('renders null', () => {
			const { container } = render(StructuredText, { props: { data: null } });

			expect(container).toMatchSnapshot();
		});
	});

	describe('simple dast /2', () => {
		const structuredText: ST = {
			value: {
				schema: 'dast',
				document: {
					type: 'root',
					children: [
						{
							type: 'heading',
							level: 1,
							children: [
								{
									type: 'span',
									value: 'This\nis a '
								},
								{
									type: 'span',
									marks: ['strong'],
									value: 'title'
								}
							]
						}
					]
				}
			}
		};

		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, { props: { data: structuredText } });

				expect(container).toMatchSnapshot();
			});
		});

		describe('with custom mark rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: structuredText, components: [[isSpan, CustomSpan]] }
				});

				expect(container).toMatchSnapshot();
			});
		});
	});

	describe('simple dast with no links/blocks', () => {
		const structuredText: ST = {
			value: {
				schema: 'dast',
				document: {
					type: 'root',
					children: [
						{
							type: 'heading',
							level: 1,
							children: [
								{
									type: 'span',
									value: 'This\nis a '
								},
								{
									type: 'span',
									marks: ['strong'],
									value: 'title'
								}
							]
						}
					]
				}
			}
		};

		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: structuredText }
				});

				expect(container).toMatchSnapshot();
			});
		});

		describe('with custom rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: { data: structuredText, components: [[isHeading, IncreasedLevelHeading]] }
				});

				expect(container).toMatchSnapshot();
			});
		});
	});

	describe('with links/blocks', () => {
		const structuredText: ST<QuoteRecord, DocPageRecord> = {
			value: {
				schema: 'dast',
				document: {
					type: 'root',
					children: [
						{
							type: 'heading',
							level: 1,
							children: [
								{
									type: 'span',
									value: 'This is a'
								},
								{
									type: 'span',
									marks: ['highlight'],
									value: 'title'
								},
								{
									type: 'inlineItem',
									item: '123'
								},
								{
									type: 'itemLink',
									item: '123',
									meta: [{ id: 'target', value: '_blank' }],
									children: [{ type: 'span', value: 'here!' }]
								}
							]
						},
						{
							type: 'block',
							item: '456'
						}
					]
				}
			},
			blocks: [
				{
					id: '456',
					__typename: 'QuoteRecord',
					quote: 'Foo bar.',
					author: 'Mark Smith'
				}
			],
			links: [
				{
					id: '123',
					__typename: 'DocPageRecord',
					title: 'How to code',
					slug: 'how-to-code'
				}
			]
		};

		describe('with default rules', () => {
			it('renders the document', () => {
				const { container } = render(StructuredText, {
					props: {
						data: structuredText,
						components: [
							[isInlineItem, InlineItem],
							[isItemLink, ItemLink],
							[isBlock, Block]
						]
					}
				});

				expect(container).toMatchSnapshot();
			});
		});

		describe('with missing component for inline records', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: { data: structuredText }
					});
				}).toThrow(RenderError);
			});
		});

		describe('with missing record', () => {
			it('raises an error', () => {
				expect(() => {
					render(StructuredText, {
						props: {
							data: { ...structuredText, links: [] },
							components: [[isInlineItem, InlineItem]]
						}
					});
				}).toThrow(RenderError);
			});
		});
	});
});
