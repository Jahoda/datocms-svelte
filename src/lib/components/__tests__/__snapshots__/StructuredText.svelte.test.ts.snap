// Vitest Snapshot v1

exports[`StructuredText > simple dast /2 > with custom mark rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <b>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;CustomSpan&gt;-->
      </b>
      
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > simple dast /2 > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > simple dast with no links/blocks > with custom rules > renders the document 1`] = `
<body>
  <div>
    <h2>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;IncreasedLevelHeading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > simple dast with no links/blocks > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with links/blocks > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This is a
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <mark>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </mark>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="/docs/how-to-code"
      >
        How to code
      </a>
      ;
      
      <!--&lt;InlineItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="/docs/how-to-code"
      >
        here!
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </a>
      
      <!--&lt;ItemLink&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <figure>
      <blockquote>
        Foo bar.
      </blockquote>
       
      <figcaption>
        Mark Smith
      </figcaption>
    </figure>
    
    <!--&lt;Block&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with no value > renders null 1`] = `
<body>
  <div>
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;
